src/pad/src/ICeWall.cpp

```diff
@@ -0,0 +1,1070 @@
+{
+  odb::dbBTerm* bterm = odb::dbBTerm::create(net, net->getConstName());
+  if (bterm == nullptr) {
+    bterm = net->get1stBTerm();
+  }
+  bterm->setSigType(net->getSigType());
+  odb::dbBPin* pin = odb::dbBPin::create(bterm);
+  odb::dbBox::create(
+      pin, layer, shape.xMin(), shape.yMin(), shape.xMax(), shape.yMax());
+  pin->setPlacementStatus(odb::dbPlacementStatus::FIRM);
+}
+
+void ICeWall::assignBump(odb::dbInst* inst, odb::dbNet* net)
+{
+  if (inst == nullptr) {
+    return;
+  }
+  if (net == nullptr) {
+    return;
+  }
+
+  odb::dbTransform xform;
+  inst->getTransform(xform);
+
+  // Connect to all iterms since this is a bump
+  for (auto* pin : inst->getITerms()) {
+    if (pin->getNet() != net) {
+      pin->connect(net);
+    }
+
+    for (auto* mpin : pin->getMTerm()->getMPins()) {
+      for (auto* geom : mpin->getGeometry()) {
+        auto* layer = geom->getTechLayer();
+        if (layer == nullptr) {
+          continue;
+        }
+
+        odb::Rect shape = geom->getBox();
+        xform.apply(shape);
+        makeBTerm(net, layer, shape);
+      }
+    }
+  }
+}
+
+void ICeWall::makeFakeSite(const std::string& name, int width, int height)
+{
+  const std::string lib_name = "FAKE_IO";
+  auto* lib = db_->findLib(lib_name.c_str());
+  if (lib == nullptr) {
+    lib = odb::dbLib::create(db_, lib_name.c_str());
+  }
+
+  auto* site = odb::dbSite::create(lib, name.c_str());
+  site->setWidth(width);
+  site->setHeight(height);
+}
+
+void ICeWall::makeIORow(odb::dbSite* horizontal_site,
+                        odb::dbSite* vertical_site,
+                        odb::dbSite* corner_site,
+                        int west_offset,
+                        int north_offset,
+                        int east_offset,
+                        int south_offset,
+                        odb::dbOrientType rotation,
+                        int ring_index)
+{
+  auto* block = getBlock();
+  if (block == nullptr) {
+    return;
+  }
+
+  if (horizontal_site == nullptr) {
+    horizontal_site = vertical_site;
+  }
+
+  if (horizontal_site == nullptr || vertical_site == nullptr
+      || corner_site == nullptr) {
+    return;
+  }
+
+  std::string name_format;
+  if (ring_index >= 0) {
+    name_format = fmt::format("IO_{{}}_{}", ring_index);
+  } else {
+    name_format = "IO_{}";
+  }
+
+  const odb::Rect die = block->getDieArea();
+
+  odb::Rect outer_io(die.xMin() + west_offset,
+                     die.yMin() + south_offset,
+                     die.xMax() - east_offset,
+                     die.yMax() - north_offset);
+
+  const int cheight = corner_site->getHeight();
+  const int cwidth
+      = std::max(vertical_site->getHeight(), corner_site->getWidth());
+
+  const int x_sites = std::ceil(static_cast<double>(outer_io.dx() - 2 * cwidth)
+                                / vertical_site->getWidth());
+  outer_io.set_xhi(outer_io.xMin() + 2 * cwidth
+                   + x_sites * vertical_site->getWidth());
```

Comment:
cwidth is a standin for the fact that the corner IO site may just be the normal IO site. I fixed the ceil, you are correct that it would grow out of bounds.